# **05. Model Training and Evaluation**
*This notebook will focus on training the machine learning model (e.g., RandomForestRegressor) and evaluating its performance using metrics like RMSLE.*

## Objectives

* Write here your notebook objective, for example, "Fetch data from Kaggle and save as raw data", or "engineer features for modelling"

## Inputs

* Write here which data or information you need to run the notebook 

## Outputs

* Write here which files, code or artefacts you generate by the end of the notebook 

## Additional Comments

* In case you have any additional comments that don't fit in the previous bullets, please state them here. 


---

# Execution Timestamp

Purpose: This code block adds a timestamp to track notebook execution
- Helps monitor when analysis was last performed
- Ensures reproducibility of results
- Useful for debugging and version control

In [1]:
# Timestamp
import datetime

import datetime
print(f"Notebook last run (end-to-end): {datetime.datetime.now()}")

Notebook last run (end-to-end): 2025-02-18 22:01:39.450206


# Project Directory Structure and Working Directory

**Purpose: This code block establishes and explains the project organization**
- Creates a standardized project structure for data science workflows
- Documents the purpose of each directory for team collaboration
- Gets current working directory for file path management

## Key Components:
1. `data/ directory` stores all datasets (raw, processed, interim)
2. `src/` contains all source code (data preparation, models, utilities)
3. `notebooks/` holds Jupyter notebooks for experimentation
4. `results/` stores output files and visualizations

## Project Root Structure

- **`data/`** - Where all your datasets live
    - `raw/` - Original, untouched data
    - `processed/` - Cleaned and prepared data
    - `interim/` - Temporary data files
- **`src/`** - Your source code
    - `data_prep/` - Code for preparing data
    - `models/` - Your ML models
    - `utils/` - Helper functions
- **`notebooks/`** - Jupyter notebooks for experiments
- **`results/`** - Model outputs and visualizations

## Setting Up Working Directory
This code block sets up the working environment by:
- Changing to the project directory where our code and data files are located
- Verifying the current working directory to ensure we're in the right place

In [2]:
import os

# Move to the desired directory
os.chdir('c:\\Users\\blign\\Dropbox\\1 PROJECT\\VS Code Project Respository\\About-BulldozerPriceGenius-_BPG-_v2')

# Get the current directory to verify the change
current_dir = os.getcwd()
current_dir

'c:\\Users\\blign\\Dropbox\\1 PROJECT\\VS Code Project Respository\\About-BulldozerPriceGenius-_BPG-_v2'

## Set Working Directory to Project Root
**Purpose: Changes the current working directory to the parent directory**
- Gets the folder one level above the current one
- Makes sure all file locations work correctly throughout the project
- Keeps files and folders organized in a clean way

In [3]:
os.chdir(os.path.dirname(current_dir))
print("You set a new current directory")

You set a new current directory


## Get Current Working Directory
**Purpose: Retrieves and stores the current working directory path**
- Gets the folder location where we're currently working
- Saves this location in a variable called current_dir so we can use it later
- Helps us find and work with files in the right place

In [4]:
import os

# Change the current working directory
os.chdir('c:\\Users\\blign\\Dropbox\\1 PROJECT\\VS Code Project Respository')

# Get the current working directory
current_dir = os.getcwd()
current_dir

'c:\\Users\\blign\\Dropbox\\1 PROJECT\\VS Code Project Respository'

---

# **Import Essential Data Science Libraries and Check Versions**

**Purpose: This code block imports fundamental Python libraries for data analysis and visualization**
- `pandas:` For data manipulation and analysis
- `numpy:` For numerical computations
- `matplotlib:` For creating visualizations and plots

**The version checks help ensure:**
- *Code compatibility across different environments*
- *Reproducibility of analysis*
- *Easy debugging of version-specific issues*


In [5]:
# Import data analysis tools
import pandas as pd
import numpy as np
import matplotlib
import matplotlib.pyplot as plt


print(f"pandas version: {pd.__version__}")
print(f"NumPy version: {np.__version__}")
print(f"matplotlib version: {matplotlib.__version__}")

pandas version: 2.2.3
NumPy version: 2.2.2
matplotlib version: 3.10.0


# **Import and Displaying the Processed Bulldozer Dataset**

This code serves three main purposes:

- Imports pandas for data manipulation
- Loads our preprocessed bulldozer dataset from a Parquet file that contains cleaned data with properly encoded categorical values and filled missing values
- Displays the first few rows of the data to verify successful loading

---

In [6]:
import pandas as pd

# Define the file path
file_path = "C:/Users/blign/Dropbox/1 PROJECT/VS Code Project Respository/About-BulldozerPriceGenius-_BPG-_v2/data/processed/TrainAndValid_object_values_as_categories_and_missing_values_filled.parquet"

# Load the Parquet file into a DataFrame
df = pd.read_parquet(file_path)

# Display the first few rows of the DataFrame
print(df.head())

   SalesID  SalePrice  MachineID  ModelID  datasource  auctioneerID  YearMade  \
0  1139246    66000.0     999089     3157         121           3.0      2004   
1  1139248    57000.0     117657       77         121           3.0      1996   
2  1139249    10000.0     434808     7009         121           3.0      2001   
3  1139251    38500.0    1026470      332         121           3.0      2001   
4  1139253    11000.0    1057373    17311         121           3.0      2007   

   MachineHoursCurrentMeter  UsageBand  fiModelDesc  ...  \
0                      68.0          2          963  ...   
1                    4640.0          2         1745  ...   
2                    2838.0          1          336  ...   
3                    3486.0          1         3716  ...   
4                     722.0          3         4261  ...   

   Undercarriage_Pad_Width_is_missing  Stick_Length_is_missing  \
0                                   1                        1   
1                   

## Loading the Preprocessed Bulldozer Dataset

This code reads our previously processed bulldozer dataset from a Parquet file. The dataset contains:

- Cleaned and properly formatted data
- Encoded categorical values
- Filled missing value

In [7]:
# Read in preprocessed dataset
df_tmp = pd.read_parquet(path="C:/Users/blign/Dropbox/1 PROJECT/VS Code Project Respository/About-BulldozerPriceGenius-_BPG-_v2/data/processed/TrainAndValid_object_values_as_categories_and_missing_values_filled.parquet",
                        engine="auto")


### Check for Missing Values

This code checks if there are any missing values in our data. It:

- Calculates the total number of missing values across all columns using pandas' isna() and sum() functions
- Provides informative feedback based on the result:
    - If no missing values are found (total = 0), confirms we can proceed with model building
    - If missing values exist, suggests reviewing our data preprocessing steps

In [8]:
# Check total number of missing values
total_missing_values = df_tmp.isna().sum().sum()

if total_missing_values == 0:
    print(f"[INFO] Total missing values: {total_missing_values} - Great! Let's build a model!")
else:
    print(f"[INFO] Uh ohh... total missing values: {total_missing_values} - Perhaps we might have to retrace our steps to fill the values?")

[INFO] Total missing values: 0 - Great! Let's build a model!


---

# Training Our Machine Learning Model

Now that we've cleaned up our data and made sure everything is in the right format, we're ready to create our price prediction model!

### Starting Small

We'll use a special type of model called a Random Forest that's good at learning patterns from data. Since we have a lot of data (over 400,000 rows), we'll first test our approach on a smaller sample of about 1,000 rows.

### Why Start Small?

Think of it like testing a recipe - it's better to try it with smaller portions first to make sure everything works before making a huge batch. This way, we can quickly fix any problems without wasting time.
 

### Setting Up Our Data

We'll organize our data into two parts:

- `Features (X)`: All the information about the bulldozers
- `Target (y)`: The actual sale prices we want to predict

### Measuring Performance

We'll use a special timing tool `(%%time)` to see how long our model takes to learn. This helps us plan for when we use the full dataset.

## Initialize and Train Random Forest Model

This code prepares and runs our model that will predict bulldozer prices.

- Imports the RandomForestRegressor class from scikit-learn's ensemble module
- Creates a model instance that utilizes all available CPU cores (n_jobs=-1)
- Trains the model using:
    - `Features (X)`: All columns except SalePrice
    - `Target (y)`: The SalePrice column we want to predict

In [9]:
from sklearn.ensemble import RandomForestRegressor

model = RandomForestRegressor(n_jobs=-1)
model.fit(X=df_tmp.drop("SalePrice", axis=1), 
        y=df_tmp.SalePrice)

RandomForestRegressor(n_jobs=-1)

### What is `n_jobs=-1`?

Setting n_jobs=-1 in a RandomForestRegressor tells the model to use all available CPU cores on your computer for training the model.

#### How it Works

Think of it like this: If your computer has 4 CPU cores, it's like having 4 workers that can process data simultaneously. When you set n_jobs=-1, you're telling the program "Use all available workers (cores)" instead of just one.

#### Benefits

- Faster training times since work is distributed across all cores
- Maximum utilization of your computer's processing power

#### Potential Drawbacks

- May slow down other programs on your computer since all cores are being used
- Could cause system stability issues on computers with limited resources

If you want to be more conservative with resource usage, you can set n_jobs to a specific number (like n_jobs=2 to use just two cores).

## Training a Sample Model for Initial Testing

This code shows how we test our Random Forest model using a small portion of our bulldozer dataset to make sure everything works correctly.

- Takes a random sample of 1,000 records from our full dataset for quick testing
- Creates a Random Forest model that uses all available CPU cores for efficient processing
- Splits the data into two parts: what we want to use to make predictions (X) and what we want to predict (y)
- Trains the model on this sample data to predict bulldozer prices

In [10]:
%%time

# Sample 1000 samples with random state 42 for reproducibility
df_tmp_sample_1k = df_tmp.sample(n=1000, random_state=42)

# Instantiate a model
model = RandomForestRegressor(n_jobs=-1) # use -1 to utilise all available processors

# Create features and labels
X_sample_1k = df_tmp_sample_1k.drop("SalePrice", axis=1) # use all columns except SalePrice as X values
y_sample_1k = df_tmp_sample_1k["SalePrice"] # use SalePrice as y values (target variable)

# Fit the model to the sample data
model.fit(X=X_sample_1k, 
          y=y_sample_1k) 

CPU times: total: 5.53 s
Wall time: 2.4 s


RandomForestRegressor(n_jobs=-1)

### Model Training Time Analysis

This output shows how long it took to train our Random Forest model on the 1,000 sample records:

- **CPU times** (total processing time across all CPU cores): 3.03 seconds
- **Wall time** (actual elapsed time): 1.5 seconds

The difference between CPU and wall time indicates effective parallel processing across multiple CPU cores.

## Evaluate Model Performance

This code checks how accurate our Random Forest model is at guessing bulldozer prices.

- Using the `score() method` to measure how well our model can guess prices by comparing its predictions to the actual bulldozer prices.
- Testing the model on the same 1,000 sample records we used for training.
- Printing the score along with the sample size for easy reference.

In [11]:
# Evaluate the model
model_sample_1k_score = model.score(X=X_sample_1k,
                                    y=y_sample_1k)

print(f"[INFO] Model score on {len(df_tmp_sample_1k)} samples: {model_sample_1k_score}")

[INFO] Model score on 1000 samples: 0.9584372165893067


### Model Performance Results

The **Random Forest model** achieved an impressive accuracy score of about `96%` when predicting bulldozer prices. Here are the key points:

- The model shows 96% accuracy in price predictions, comparable to getting 96 out of 100 questions correct.
- Testing was conducted on a limited sample:
    - Only 1,000 bulldozers were used.
    - All available CPU power was utilized for efficient processing.
- Practical implications:
    - Provides reliable price estimates for buyers and sellers.
    - Results may vary when applied to the complete dataset.

## Training the Full Random Forest Model

This code runs our Random Forest model using all our bulldozer data to make better price predictions.

- Measures the training time using a special timer command `*%%*time`
- Uses all available computer processing power to run the model faster
- Gets the data ready by:
    - Taking out the price information (SalePrice) from the main dataset
    - Keeping the price information separate to use as our target values
- Uses all available bulldozer data to help the model learn how to predict prices accurately

In [12]:
%%time

# Instantiate model
model = RandomForestRegressor(n_jobs=-1) # note: this could take quite a while depending on your machine

# Create features and labels with entire dataset
X_all = df_tmp.drop("SalePrice", axis=1)
y_all = df_tmp["SalePrice"]

# Fit the model
model.fit(X=X_all, 
        y=y_all)

CPU times: total: 32min 3s
Wall time: 5min 41s


RandomForestRegressor(n_jobs=-1)

## Evaluating Model Performance on Full Dataset

This code block evaluates how well our Random Forest model performs on the complete bulldozer dataset. It:

- Uses the score() method to measure prediction accuracy
- Calculates performance using both features (X_all) and actual prices (y_all)
- Prints the final score along with the total number of samples used

This evaluation is important because it shows us how well our model can predict bulldozer prices when using the entire dataset, rather than just the small sample we tested earlier.

In [13]:
# Evaluate the model
model_sample_all_score = model.score(X=X_all,
                                     y=y_all)

print(f"[INFO] Model score on {len(df_tmp)} samples: {model_sample_all_score}")

[INFO] Model score on 412698 samples: 0.9875748328649882


---

# **Understanding Data Splitting: A Simple Guide**

When working with machine learning projects, it's crucial to split your data properly. Here's why and how we do it:

### What is Data Splitting?

Think of data splitting like dividing a recipe book into three parts:

- `Training data`: The recipes you practice with
- `Validation data`: The recipes you test yourself on
- `Test data`: The final exam recipes

### Why Time Matters

For projects involving time-based predictions (like our bulldozer price predictions), we need to be extra careful with how we split the data. Random splitting won't work because it mixes up the timeline.

### How We Split the Data

In this project, we organize our data by dates:

- `Training data`: Everything up until 2011
- `Validation data`: January 1 to April 30, 2012
- `Testing data`: May 1 to November 2012

This approach ensures we're training our model on past data to predict future prices, just like how we'd use historical prices to guess future ones in the real world.

In [14]:
# Import train samples (making sure to parse dates and then sort by them)
train_df = pd.read_csv(filepath_or_buffer="C:/Users/blign/Dropbox/1 PROJECT/VS Code Project Respository/About-BulldozerPriceGenius-_BPG-_v2/data/raw/bluebook-for-bulldozers/Train.csv",
                       parse_dates=["saledate"],
                       low_memory=False).sort_values(by="saledate", ascending=True)

# Import validation samples (making sure to parse dates and then sort by them)
valid_df = pd.read_csv(filepath_or_buffer="C:/Users/blign/Dropbox/1 PROJECT/VS Code Project Respository/About-BulldozerPriceGenius-_BPG-_v2/data/raw/bluebook-for-bulldozers/Valid.csv",
                       parse_dates=["saledate"])

# The ValidSolution.csv contains the SalePrice values for the samples in Valid.csv
valid_solution = pd.read_csv(filepath_or_buffer="C:/Users/blign/Dropbox/1 PROJECT/VS Code Project Respository/About-BulldozerPriceGenius-_BPG-_v2/data/raw/bluebook-for-bulldozers/ValidSolution.csv")

# Map valid_solution to valid_df
valid_df["SalePrice"] = valid_df["SalesID"].map(valid_solution.set_index("SalesID")["SalePrice"])

# Make sure valid_df is sorted by saledate still
valid_df = valid_df.sort_values("saledate", ascending=True).reset_index(drop=True)

# How many samples are in each DataFrame?
print(f"[INFO] Number of samples in training DataFrame: {len(train_df)}")
print(f"[INFO] Number of samples in validation DataFrame: {len(valid_df)}")

[INFO] Number of samples in training DataFrame: 401125
[INFO] Number of samples in validation DataFrame: 11573


## Loading and Processing Training/Validation Data

This code prepares our bulldozer dataset for machine learning. Here's what it does in simple terms:

- Loads the training data from a file and makes sure all dates are in the right format and order
- Gets the test data ready the same way
- Adds the actual bulldozer prices to the test data
- Makes sure all the data is arranged by date, which helps us make better predictions
- Shows us how many bulldozers we have in our training and test sets

## Examining `Training Data` Sample

This code displays a random sample of 5 rows from our training dataset. This helps us:

- Verify the data is loaded correctly.
- Understand what features we're working with.
- Spot any potential issues in the data structure.

In [15]:
# Let's check out the training DataFrame
train_df.sample(5)

SalesID  SalePrice  MachineID  ModelID  datasource  auctioneerID  \
310063  2268130      31000    1432451     3350         136           8.0   
294293  2221073      38500     113246     1177         136          12.0   
334387  2346854      13500    1703168    14286         136           NaN   
36460   1273453      11500     569896     3269         132           1.0   
231904  1726331      27500    1141676      251         132           1.0   

        YearMade  MachineHoursCurrentMeter UsageBand   saledate  ...  \
310063      1976                       0.0       NaN 2007-09-12  ...   
294293      1992                       0.0       NaN 2008-11-19  ...   
334387      2004                     980.0       Low 2009-03-11  ...   
36460       1999                       NaN       NaN 2006-10-10  ...   
231904      1987                       NaN       NaN 1994-12-09  ...   

       Undercarriage_Pad_Width         Stick_Length                Thumb  \
310063                     NaN                  NaN                  NaN   
294293                 32 inch  None or Unspecified  None or Unspecified   
334387                     NaN                  NaN                  NaN   
36460                      NaN                  NaN                  NaN   
231904                     NaN                  NaN                  NaN   

            Pattern_Changer Grouser_Type     Backhoe_Mounting Blade_Type  \
310063                  NaN          NaN                  NaN        NaN   
294293  None or Unspecified       Triple                  NaN        NaN   
334387                  NaN          NaN                  NaN        NaN   
36460                   NaN          NaN                  NaN        NaN   
231904                  NaN          NaN  None or Unspecified   Straight   

            Travel_Controls Differential_Type Steering_Controls  
310063                  NaN               NaN               NaN  
294293                  NaN               NaN               NaN  
334387                  NaN               NaN               NaN  
36460                   NaN               NaN               NaN  
231904  None or Unspecified               NaN               NaN  

[5 rows x 53 columns]

## Examining `Validation Data` Sample

This code displays a random sample of 5 rows from our validation dataset to:

- Verify the validation data is structured correctly.
- Compare validation data features with training data.
- Check for any inconsistencies between datasets.

In [16]:
# And how about the validation DataFrame?
valid_df.sample(5)

SalesID  MachineID  ModelID  datasource  auctioneerID  YearMade  \
8579   4252766    2277101    13247         172             1      2006   
11106  6283175    1888992     1893         149             1      1999   
1764   4394396    2266732     5874         172             1      2006   
2239   4299051    2312759     1580         172             1      2003   
8081   4268618    1940508     3824         172             1      2001   

       MachineHoursCurrentMeter UsageBand   saledate fiModelDesc  ...  \
8579                     1626.0       Low 2012-03-28      580MII  ...   
11106                       NaN       NaN 2012-04-24     650HLGP  ...   
1764                    10648.0      High 2012-02-07       L220E  ...   
2239                     3309.0       Low 2012-02-12      D5GLGP  ...   
8081                     9462.0    Medium 2012-03-26        950G  ...   

      Stick_Length Thumb Pattern_Changer Grouser_Type     Backhoe_Mounting  \
8579           NaN   NaN             NaN          NaN                  NaN   
11106          NaN   NaN             NaN          NaN  None or Unspecified   
1764           NaN   NaN             NaN          NaN                  NaN   
2239           NaN   NaN             NaN          NaN  None or Unspecified   
8081           NaN   NaN             NaN          NaN                  NaN   

      Blade_Type      Travel_Controls Differential_Type Steering_Controls  \
8579         NaN                  NaN               NaN               NaN   
11106        PAT  None or Unspecified               NaN               NaN   
1764         NaN                  NaN          Standard      Conventional   
2239         PAT  None or Unspecified               NaN               NaN   
8081         NaN                  NaN          Standard      Conventional   

      SalePrice  
8579    32000.0  
11106   24000.0  
1764    70000.0  
2239    42000.0  
8081    72500.0  

[5 rows x 53 columns]

# Creating Time-Based Features from Sale Dates

Here's what this code does with sale dates to help predict bulldozer prices better:

- **Why we need it:**
    - Helps spot patterns in when bulldozers sell for more or less money.
    - Makes it easier to see how prices change over time.
- **What it does:**
    - Breaks down each sale date into useful parts (year, month, and day).
    - Adds helpful details like which day of the week and what time of year it was.
    - Removes the original date to keep things simple.

In [17]:
# Make a function to add date columns
def add_datetime_features_to_df(df, date_column="saledate"):
    # Add datetime parameters for saledate
    df["saleYear"] = df[date_column].dt.year
    df["saleMonth"] = df[date_column].dt.month
    df["saleDay"] = df[date_column].dt.day
    df["saleDayofweek"] = df[date_column].dt.dayofweek
    df["saleDayofyear"] = df[date_column].dt.dayofyear

    # Drop original saledate column
    df.drop("saledate", axis=1, inplace=True)

    return df

train_df = add_datetime_features_to_df(df=train_df)
valid_df = add_datetime_features_to_df(df=valid_df)

## Viewing Time-Based Features

This code displays:

- A random sample of 5 rows showing our new time features.
- Includes year, month, day, day of week, and day of year.
- Helps verify:
    - Date breakdowns are correct.
    - Temporal data structure is proper.

In [18]:
# Display the last 5 columns (the recently added datetime breakdowns)
train_df.iloc[:, -5:].sample(5)

saleYear  saleMonth  saleDay  saleDayofweek  saleDayofyear
241408      2003         12        6              5            340
8784        2005         11        3              3            307
11487       2008          1       31              3             31
340261      2007         11       16              4            320
93213       1991         10        3              3            276

---

# **Trying To Fit A Model On Our `Training Data`**
Let's start by testing our model right away with our data. This approach helps us understand what we're working with quickly.

When we test the model, one of two things will happen:

- If it works: We'll look at the results to see how well it did
- If it doesn't work: We'll learn what we need to fix in our data

To get started, we'll:

- Split our data into two parts:
    - The features (X): All the bulldozer information except the price
    - The target (y): Just the price we want to predict

Then we'll use a special tool called RandomForestRegressor to make price predictions using our training data.

In [ ]:
# Split training data into features and labels
# X_train = train_df.drop("SalePrice", axis=1)
# y_train = train_df["SalePrice"]

# Split validation data into features and labels
# X_valid = valid_df.drop("SalePrice", axis=1)
# y_valid = valid_df["SalePrice"]

# Create a model
# model = RandomForestRegressor(n_jobs=-1)

# Fit a model to the training data only
# model.fit(X=X_train,
#          y=y_train)

## Handling Data Type Mismatch in Model Training

We've hit a small roadblock in our model training. Our program is having trouble understanding some of the data because it's expecting numbers but found text values (like the word 'Medium') instead.

Here's what's happening:

- Our model can only work with numbers.
- We found text values in our data (like 'Medium').
- This happened because we loaded our raw data file (`Train.csv`) instead of using our processed version.

The good news is this is a common issue and we know exactly how to fix it!

---

# **Encoding categorical features as numbers using Scikit-Learn**

## Identifying Numerical and Categorical Features

Our code looks at our data and sorts it into two simple groups:

- **Numbers**: Things we can count or measure (like prices and years).
- **Categories**: Words or labels that describe things (like model names or condition types).

This sorting is important because:

- Our computer needs to handle numbers and words differently.
- It helps us prepare our data the right way.
- It makes sure our computer can understand and use all our information.

In [ ]:
# Import necessary libraries
import pandas as pd
from sklearn.model_selection import train_test_split

# Load the dataset
df_tmp = pd.read_csv("data/raw/bluebook-for-bulldozers/TrainAndValid.csv",
                     low_memory=False,
                     parse_dates=["saledate"])

# Feature engineering: Add datetime parameters and drop original saledate column
df_tmp["saleYear"] = df_tmp.saledate.dt.year
df_tmp["saleMonth"] = df_tmp.saledate.dt.month
df_tmp["saleDay"] = df_tmp.saledate.dt.day
df_tmp["saleDayofweek"] = df_tmp.saledate.dt.dayofweek
df_tmp["saleDayofyear"] = df_tmp.saledate.dt.dayofyear
df_tmp.drop("saledate", axis=1, inplace=True)

# Convert object type columns to category
for label, content in df_tmp.items():
    if pd.api.types.is_object_dtype(content):
        df_tmp[label] = df_tmp[label].astype("category")

# Split data into training and testing sets
X = df_tmp.drop("SalePrice", axis=1)
y = df_tmp.SalePrice
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)  # Adjust test_size and random_state as needed

# Define numerical and categorical features
numerical_features = [label for label, content in X_train.items() if pd.api.types.is_numeric_dtype(content)]
categorical_features = [label for label, content in X_train.items() if not pd.api.types.is_numeric_dtype(content)]

print(f"[INFO] Numeric features: {numerical_features}")
print(f"[INFO] Categorical features: {categorical_features}")


[INFO] Numeric features: ['SalesID', 'MachineID', 'ModelID', 'datasource', 'auctioneerID', 'YearMade', 'MachineHoursCurrentMeter', 'saleYear', 'saleMonth', 'saleDay', 'saleDayofweek', 'saleDayofyear']
[INFO] Categorical features: ['UsageBand', 'fiModelDesc', 'fiBaseModel', 'fiSecondaryDesc', 'fiModelSeries', 'fiModelDescriptor', 'ProductSize', 'fiProductClassDesc', 'state', 'ProductGroup', 'ProductGroupDesc', 'Drive_System', 'Enclosure', 'Forks', 'Pad_Type', 'Ride_Control', 'Stick', 'Transmission', 'Turbocharged', 'Blade_Extension', 'Blade_Width', 'Enclosure_Type', 'Engine_Horsepower', 'Hydraulics', 'Pushblock', 'Ripper', 'Scarifier', 'Tip_Control', 'Tire_Size', 'Coupler', 'Coupler_System', 'Grouser_Tracks', 'Hydraulics_Flow', 'Track_Type', 'Undercarriage_Pad_Width', 'Stick_Length', 'Thumb', 'Pattern_Changer', 'Grouser_Type', 'Backhoe_Mounting', 'Blade_Type', 'Travel_Controls', 'Differential_Type', 'Steering_Controls']


## Converting Categorical Features to Numbers

Our code takes information that's written as text and turns it into numbers so our computer can work with it. 
**Here's why we need to do this:**

- Computers can only understand and work with numbers.
- Some of our information, like equipment models and conditions, is written as text.
- We need to change this text into numbers, but keep its original meaning.

**The code does these main things:**

- Uses a special tool to change text into numbers.
- Handles any new or unexpected text by marking it as missing.
- Makes these changes to both our training and testing data.

In [36]:
from sklearn.preprocessing import OrdinalEncoder
from sklearn.model_selection import train_test_split
import numpy as np

# Split data into training and validation sets
X_train, X_valid, y_train, y_valid = train_test_split(
    df_tmp.drop("SalePrice", axis=1), df_tmp["SalePrice"], test_size=0.2, shuffle=False
)

# Get a list of categorical features
categorical_features = X_train.select_dtypes(include=['category', object]).columns.tolist()

# 1. Create an ordinal encoder 
ordinal_encoder = OrdinalEncoder(categories="auto",
                                 handle_unknown="use_encoded_value",
                                 unknown_value=np.nan,
                                 encoded_missing_value=np.nan) 

# 2. Fit and transform the categorical columns of X_train
X_train_preprocessed = X_train.copy() 
X_train_preprocessed[categorical_features] = ordinal_encoder.fit_transform(X_train_preprocessed[categorical_features].astype(str))

# 3. Transform the categorical columns of X_valid 
X_valid_preprocessed = X_valid.copy()
X_valid_preprocessed[categorical_features] = ordinal_encoder.transform(X_valid_preprocessed[categorical_features].astype(str))

## Display First Few Rows of Training Data

This code displays the first 5 rows of our training dataset (`X_train`) to:

- Verify our data is structured correctly
- Preview the features we'll use for training
- Ensure all columns are present and properly formatted


In [39]:
X_train.head()

SalesID  MachineID  ModelID  datasource  auctioneerID  YearMade  \
0  1139246     999089     3157         121           3.0      2004   
1  1139248     117657       77         121           3.0      1996   
2  1139249     434808     7009         121           3.0      2001   
3  1139251    1026470      332         121           3.0      2001   
4  1139253    1057373    17311         121           3.0      2007   

   MachineHoursCurrentMeter UsageBand fiModelDesc fiBaseModel  ...  \
0                      68.0       Low        521D         521  ...   
1                    4640.0       Low      950FII         950  ...   
2                    2838.0      High         226         226  ...   
3                    3486.0      High    PC120-6E       PC120  ...   
4                     722.0    Medium        S175        S175  ...   

  Backhoe_Mounting Blade_Type Travel_Controls Differential_Type  \
0              NaN        NaN             NaN          Standard   
1              NaN        NaN             NaN          Standard   
2              NaN        NaN             NaN               NaN   
3              NaN        NaN             NaN               NaN   
4              NaN        NaN             NaN               NaN   

  Steering_Controls saleYear saleMonth saleDay saleDayofweek saleDayofyear  
0      Conventional     2006        11      16             3           320  
1      Conventional     2004         3      26             4            86  
2               NaN     2004         2      26             3            57  
3               NaN     2011         5      19             3           139  
4               NaN     2009         7      23             3           204  

[5 rows x 56 columns]

## Display Preprocessed Training Data

This code displays the first 5 rows of our preprocessed training dataset to:

- Verify our categorical features have been successfully converted to numerical values
- Confirm the preprocessing steps were applied correctly
- Check the data format is now suitable for model training

In [40]:
X_train_preprocessed.head()

SalesID  MachineID  ModelID  datasource  auctioneerID  YearMade  \
0  1139246     999089     3157         121           3.0      2004   
1  1139248     117657       77         121           3.0      1996   
2  1139249     434808     7009         121           3.0      2001   
3  1139251    1026470      332         121           3.0      2001   
4  1139253    1057373    17311         121           3.0      2007   

   MachineHoursCurrentMeter  UsageBand  fiModelDesc  fiBaseModel  ...  \
0                      68.0        1.0        868.0        290.0  ...   
1                    4640.0        1.0       1577.0        507.0  ...   
2                    2838.0        0.0        299.0        108.0  ...   
3                    3486.0        0.0       3213.0       1295.0  ...   
4                     722.0        2.0       3633.0       1430.0  ...   

   Backhoe_Mounting  Blade_Type  Travel_Controls  Differential_Type  \
0               2.0        10.0              7.0                3.0   
1               2.0        10.0              7.0                3.0   
2               2.0        10.0              7.0                4.0   
3               2.0        10.0              7.0                4.0   
4               2.0        10.0              7.0                4.0   

   Steering_Controls  saleYear  saleMonth  saleDay  saleDayofweek  \
0                1.0      2006         11       16              3   
1                1.0      2004          3       26              4   
2                5.0      2004          2       26              3   
3                5.0      2011          5       19              3   
4                5.0      2009          7       23              3   

   saleDayofyear  
0            320  
1             86  
2             57  
3            139  
4            204  

[5 rows x 56 columns]

## Finding Missing Data in Categories

This code helps us find where data is missing in our categories. It does these simple things:

- Looks at just the category columns in our training data
- Counts how many empty spots we have in each column
- Lists them in order, showing which columns have the most missing data first
- Shows us the top 10 problem areasThe simpler rewrite:
    - Uses shorter, clearer sentences
    - Removes technical terms like "categorical features" and "NA values"
    - Maintains the same structure and information
    - Makes the purpose more immediately clear
    - Uses more everyday language while keeping the meaning intact

In [44]:
X_train[categorical_features].isna().sum().sort_values(ascending=False)[:10]

Tip_Control          308533
Enclosure_Type       308533
Engine_Horsepower    308533
Blade_Extension      308533
Blade_Width          308533
Pushblock            308533
Scarifier            308523
Grouser_Tracks       296496
Hydraulics_Flow      296496
Coupler_System       296411
dtype: int64

## Examining Category Encodings

This code shows us the first three types of data we've converted into numbers, helping us:

- Check if our text labels were properly turned into numbers.
- Look at how each label was matched with its number.
- Make sure we kept the right connections between related labels.

In [45]:
# Let's inspect the first three categories
ordinal_encoder.categories_[:3]

[array(['High', 'Low', 'Medium', np.str_('nan')], dtype=object),
 array(['100C', '104', '1066', ..., 'ZX800', 'ZX800LC', 'ZX850H'],
       shape=(4312,), dtype=object),
 array(['10', '100', '104', ..., 'ZX80', 'ZX800', 'ZX850'],
       shape=(1824,), dtype=object)]

## Creating Category-to-Number Mapping Dictionary

This code helps us keep track of how we change text labels into numbers by creating an easy-to-use reference list. It works like a translator:

- Creates a dictionary that maps each category column to its numerical values.
- Preserves the relationship between original text values and their numeric codes.
- Helps us translate numbers back to their original categories when needed.

In [46]:
# Create a dictionary of dictionaries mapping column names and their variables to their numerical encoding
column_to_category_mapping = {}

for column_name, category_values in zip(categorical_features, ordinal_encoder.categories_):
    int_to_category = {i: category for i, category in enumerate(category_values)}
    column_to_category_mapping[column_name] = int_to_category

# Inspect an example column name to category mapping
column_to_category_mapping["UsageBand"]

{0: 'High', 1: 'Low', 2: 'Medium', 3: np.str_('nan')}

## Converting Numbers Back to Categories

This code takes our number data and changes it back to the original words and labels so we can check if everything was converted correctly.

- Creates a copy of our preprocessed data to preserve the original.
- Uses `inverse_transform` to convert numbers back to their original categories.
- Organizes the data into a readable DataFrame format.
- Shows a random sample of 5 rows to verify the conversion worked correctly.

In [48]:
# Create a copy of the preprocessed DataFrame
X_train_unprocessed = X_train_preprocessed[categorical_features].copy()

# This will return an array of the original untransformed data
X_train_unprocessed = ordinal_encoder.inverse_transform(X_train_unprocessed)

# Turn back into a DataFrame for viewing pleasure
X_train_unprocessed_df = pd.DataFrame(X_train_unprocessed, columns=categorical_features)

# Check out a sample
X_train_unprocessed_df.sample(5)

UsageBand fiModelDesc fiBaseModel fiSecondaryDesc fiModelSeries  \
15621        Low      SK75UR        SK75              UR           nan   
14557     Medium        420D         420               D           nan   
70789        nan       IT28F        IT28               F           nan   
292434       nan     580MTII         580              MT            II   
293084       nan         D6D          D6               D           nan   

       fiModelDescriptor ProductSize  \
15621                nan        Mini   
14557                nan         nan   
70789                nan         nan   
292434               nan         nan   
293084               nan      Medium   

                                       fiProductClassDesc           state  \
15621   Hydraulic Excavator, Track - 6.0 to 8.0 Metric...         Alabama   
14557   Backhoe Loader - 14.0 to 15.0 Ft Standard Digg...           Idaho   
70789            Wheel Loader - 120.0 to 135.0 Horsepower           Texas   
292434  Backhoe Loader - 14.0 to 15.0 Ft Standard Digg...         Florida   
293084  Track Type Tractor, Dozer - 130.0 to 160.0 Hor...  South Carolina   

       ProductGroup  ... Undercarriage_Pad_Width         Stick_Length  \
15621           TEX  ...     None or Unspecified  None or Unspecified   
14557            BL  ...                     nan                  nan   
70789            WL  ...                     nan                  nan   
292434           BL  ...                     nan                  nan   
293084          TTT  ...                     nan                  nan   

                      Thumb      Pattern_Changer Grouser_Type  \
15621   None or Unspecified  None or Unspecified       Double   
14557                   nan                  nan          nan   
70789                   nan                  nan          nan   
292434                  nan                  nan          nan   
293084                  nan                  nan          nan   

           Backhoe_Mounting Blade_Type      Travel_Controls Differential_Type  \
15621                   nan        nan                  nan               nan   
14557                   nan        nan                  nan               nan   
70789                   nan        nan                  nan          Standard   
292434                  nan        nan                  nan               nan   
293084  None or Unspecified   Straight  None or Unspecified               nan   

       Steering_Controls  
15621                nan  
14557                nan  
70789       Conventional  
292434               nan  
293084               nan  

[5 rows x 44 columns]

# **Fitting A Model To Our Preprocessed Training Data** 

### Getting Our Model Ready

Now that we've organized our data, we can start training our computer to make predictions. 

## Training Our Random Forest Model

This code gets our bulldozer price prediction system ready to work. It:

- Uses `%%time` to measure how long the training takes.
- Creates a RandomForestRegressor with parallel processing `(n_jobs=-1)`.
- Trains the model using our prepared features (X_train_preprocessed) and target values (y_train).

In [49]:
%%time

# Instantiate a Random Forest Regression model
model = RandomForestRegressor(n_jobs=-1)

# Fit the model to the preprocessed training data
model.fit(X=X_train_preprocessed,
          y=y_train)

CPU times: total: 22min 28s
Wall time: 4min 8s


RandomForestRegressor(n_jobs=-1)

## Evaluating Model Performance on Validation Set

This code helps us check if our model is working well by testing it on new data it hasn't seen before (the validation set). We can:

- See how long the testing takes using the %%time command.
- Get a score between 0 and 1 that shows how accurate our predictions are.
- The closer the score is to 1, the better our model is at predicting bulldozer prices.

In [ ]:
%%time

# Check model performance on the validation set
# model.score(X=X_valid,
#            y=y_valid)

ValueError: could not convert string to float: 'Medium'

### Error Explanation: Unable to Process Text Data
Here are the bullet points from the error explanation:

- The model is trying to work with text data (specifically the word 'Medium') but it can only handle numerical values
- This is like trying to do math with words instead of numbers - it just doesn't work
- The validation data needs to be preprocessed (converted to numbers) in the same way as the training data before the model can use it

In [51]:
%%time

# Check model performance on the validation set
model.score(X=X_valid_preprocessed,
            y=y_valid)

CPU times: total: 12.5 s
Wall time: 6.66 s


0.7837103481203419

### Model Performance Results

Our bulldozer price prediction model shows:

- Achieved a score of 0.784 (78.4%) on new data.
- Demonstrates strong real-world prediction capability.
- Provides reliable estimates for future bulldozer pricing.

#### Why This Score Matters

- The model achieved a score of 78.4% on new, unseen data.
- While an earlier test showed a higher score of 98.75%, this was less reliable since it was tested on data the model had already learned from.
- **The `78.4%` score is a more accurate representation of how the model will likely perform in real-world predictions.**

#### Real-World Implications

- The model achieved a score of 78.4% on new, unseen data.
- While an earlier test showed 98.75%, this higher score was less reliable since it came from data the model had already seen.
- The 78.4% score better represents how the model will likely perform in real-world predictions.

## Checking Training Set Performance

This code shows us how well the model learned from its training data:

- Uses %%time to track how long the evaluation takes
- Applies model.score() to compare predicted prices against actual training data prices
- Helps us understand if the model learned the training patterns effectively

In [52]:
%%time

# Check model performance on the training set
model.score(X=X_train_preprocessed,
            y=y_train)

CPU times: total: 33.8 s
Wall time: 20.1 s


0.9872146214280765

### Model Performance Results

The machine learning model for predicting bulldozer prices showed impressive results in two key areas:

##### Training Set Performance

Based on the selection, here are the key points about the model's training performance:

- **The model achieved a score of `98.72%` when tested on training data.**
- This high score indicates the model successfully learned the patterns in the training data.
- The model performed better on training data than validation data, which is expected and normal in machine learning.

# **Building an evaluation function**

To ensure our machine learning model performs well, we need a way to measure its accuracy. We'll create an evaluation function that helps us:

- Compare predicted prices against actual prices
- Track model performance consistently across different tests
- Use industry-standard metrics for bulldozer price prediction

### Key Evaluation Metrics

We'll use two main metrics to evaluate our model:

### 1. Root Mean Squared Log Error (RMSLE)

This is the official metric used in the Kaggle Bulldozer competition. It measures relative errors rather than absolute ones, meaning:

- A $100 error on a $1,000 prediction (10%) is considered worse than
- A $100 error on a $10,000 prediction (1%)

### 2. Mean Absolute Error (MAE)

This gives us a different perspective by measuring absolute differences between predictions and actual values.

### Implementation Details

Our evaluation function will calculate:

- MAE (Mean Absolute Error) - lower is better
- RMSLE (Root Mean Squared Log Error) - lower is better
- R² Score (Coefficient of Determination) - higher is better

We'll use scikit-learn's built-in metrics and the model's predict() method to generate these scores.

In [53]:
# Create evaluation function (the competition uses Root Mean Square Log Error)
from sklearn.metrics import mean_absolute_error, root_mean_squared_log_error

# Create function to evaluate our model
def show_scores(model, 
                train_features=X_train_preprocessed,
                train_labels=y_train,
                valid_features=X_valid_preprocessed,
                valid_labels=y_valid):
    
    # Make predictions on train and validation features
    train_preds = model.predict(X=train_features)
    val_preds = model.predict(X=valid_features)

    # Create a scores dictionary of different evaluation metrics
    scores = {"Training MAE": mean_absolute_error(y_true=train_labels, 
                                                  y_pred=train_preds),
              "Valid MAE": mean_absolute_error(y_true=valid_labels, 
                                               y_pred=val_preds),
              "Training RMSLE": root_mean_squared_log_error(y_true=train_labels, 
                                                            y_pred=train_preds),
              "Valid RMSLE": root_mean_squared_log_error(y_true=valid_labels, 
                                                         y_pred=val_preds),
              "Training R^2": model.score(X=train_features, 
                                          y=train_labels),
              "Valid R^2": model.score(X=valid_features, 
                                       y=valid_labels)}
    return scores

## Evaluating Model Performance

This code section serves to test our evaluation function and display comprehensive model performance metrics. It:

- Calls our custom show_scores() function with our trained model.
- Stores the results in model_scores variable.
- Displays various metrics including MAE, RMSLE, and R² scores for both training and validation data.

In [54]:
# Try our model scoring function out
model_scores = show_scores(model=model)
model_scores

{'Training MAE': 1602.9096797290997,
 'Valid MAE': 7339.493018899927,
 'Training RMSLE': 0.08527066465584583,
 'Valid RMSLE': 0.3436579582939107,
 'Training R^2': 0.9872146214280765,
 'Valid R^2': 0.7837103481203419}

### Model Performance Results

##### Training Data Results (How well it learned)

- Price predictions were off by about $1,600 on average.
- The model was 98.7% accurate on data it trained with.
- Very small error rate of 0.085 (closer to 0 is better).

##### Real-World Performance (New Data)

- Price predictions were off by about $7,300 on average.
- The model was 78.4% accurate on new data.
- Higher error rate of 0.344 (expected for new data).

##### What This Means

- The model learned its training data very well (98.7% accuracy).
- When faced with new data, it's still quite good (78.4% accuracy).
- This difference is normal - models usually perform better on data they've seen before.

# **Tuning our model's hyperparameters**